src/main/kotlin/org/rust/ide/typing/paste/JsonStructParser.kt

```diff
@@ -0,0 +1,140 @@
+    data class Array(val type: DataType) : DataType()
+}
+
+data class Field(val key: String, val type: DataType)
+
+class Struct(val fields: List<Field>) {
+    // Consider structs with the same field names and types to be the same
+    // The fields are sorted before comparison, so two struct with the same field types and names, but
+    // with a different field order, are considered to be the same struct.
+    override fun equals(other: Any?): Boolean {
+        if (other == null) return false
+        if (other !is Struct) return false
+
+        val selfFields = fields.sortedBy { it.key }
+        val otherFields = other.fields.sortedBy { it.key }
+        return selfFields == otherFields
+    }
+
+    override fun hashCode(): Int = fields.sortedBy { it.key }.hashCode()
+}
+
+/**
+ * Try to parse the input text as a JSON object.
+ *
+ * Extract a list of (unique) structs that were encountered in the JSON object.
+ */
+fun extractStructsFromJson(text: String): List<Struct>? {
+    return try {
+        val factory = JsonFactory()
+        val parser = factory.createParser(text)
+
+        val structParser = StructParser()
+        parser.use {
+            // There must be an object at the root
+            val token = parser.nextToken()
+            if (token != JsonToken.START_OBJECT) return null
+
+            structParser.parseStruct(parser)
+        }
+
+        structParser.encounteredStructs
+    } catch (e: IOException) {
+        null
+    }
+}
+
+private class StructParser {
+    private val structMap = mutableSetOf<Struct>()
+    private val structs = mutableListOf<Struct>()
+
+    val encounteredStructs: List<Struct> = structs
+
+    fun parseStruct(parser: JsonParser): Struct? {
+        if (parser.currentToken != JsonToken.START_OBJECT) return null
+
+        val fields = mutableListOf<Field>()
+        val fieldSet = mutableSetOf<String>()
+        while (true) {
+            val field = parseField(parser) ?: break
+
+            // Ignore duplicate fields
+            if (field.key in fieldSet) continue
+
+            fieldSet.add(field.key)
+            fields.add(field)
+        }
+
+        if (parser.currentToken != JsonToken.END_OBJECT) return null
+
+        val struct = Struct(fields)
+        if (struct !in structMap) {
+            structMap.add(struct)
+            structs.add(struct)
+        }
+        return struct
+    }
+
+    private fun parseField(parser: JsonParser): Field? {
+        if (parser.nextToken() != JsonToken.FIELD_NAME) return null
+
+        val name = parser.currentName
+        val type = parseDataType(parser)
+        return Field(name, type)
+    }
+
+    private fun parseArray(parser: JsonParser): DataType? {
+        if (parser.currentToken != JsonToken.START_ARRAY) return null
+
+        // Use the first array data type to set the data type of the whole array
+        // If the array is empty, this will return DataType.Unknown
+        val dataType = parseDataType(parser)
```

Comment:
Consider using types of all array elements to support cases like `[null, 1, 2]` → `Vec<Option<i64>>`

Reply:
What should happen in the case of mixed arrays?
1) Empty array or only nulls -> unknown type
2) Single type -> `Type`
3) Single type with null(s) -> `Option<Type>`
4) Multiple types -> Take first type?

Reply:
> 1. Empty array or only nulls -> unknown type
Empty array → unknown type
Array of nulls → `Option<_>`

> 2. Single type -> `Type`
> 3. Single type with null(s) -> `Option<Type>`
👍

> 4. Multiple types -> Take first type?
I would use unknown type